File: code/ex130.cpp
const int dir[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

class Solution {
    int m, n;
public:
    void solve(vector<vector<char>>& board) {
        m = board.size();    if (m == 0)    return;
        n = board[0].size();    if (n == 0)    return;
        for (int i = 0; i < m; i++) {
            if (board[i][0] == 'O')    dfs(board, i, 0);
            if (board[i][n - 1] == 'O')    dfs(board, i, n - 1);
        }
        for (int j = 0; j < n; j++) {
            if (board[0][j] == 'O')    dfs(board, 0, j);
            if (board[m - 1][j] == 'O')    dfs(board, m - 1, j);
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'O')    board[i][j] = 'X';
                if (board[i][j] == '!')    board[i][j] = 'O';
            }
        }
    }
    
private:
    void dfs(vector<vector<char>>& board, int i, int j) {
        if (i < 0 || j < 0 || i >= m || j >= n || board[i][j] != 'O')    return;
        board[i][j] = '!';
        for (auto& d : dir) {
            int ni = i + d[0], nj = j + d[1];
            dfs(board, ni, nj);
        }
    }
};

File: code/ex130.java
class Solution {
    public int m, n;
    
    public void solve(char[][] board) {
        m = board.length;    if (m == 0)    return;
        n = board[0].length;    if (n == 0)    return;
        
        boolean[][] notFlip = new boolean[m][n];
        
        // DFS through the top and bottom row
        for (int i = 0; i < n; i++) {
            if (board[0][i] == 'O')    dfs(board, notFlip, 0, i);
            if (board[m - 1][i] == 'O')    dfs(board, notFlip, m - 1, i);
        }
        
        // DFS through the left most and right most column
        for (int i = 1; i < m - 1; i++) {
            if (board[i][0] == 'O')    dfs(board, notFlip, i, 0);
            if (board[i][n - 1] == 'O')    dfs(board, notFlip, i, n - 1);
        }
        
        // Flip inner 'O's
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!notFlip[i][j])    board[i][j] = 'X';
            }
        }
    }
    
    private void dfs(char[][] board, boolean[][] notFlip, int i, int j) {
        notFlip[i][j] = true;
        
        // Go four directions
        
        // Go down
        if (i + 1 < m && !notFlip[i + 1][j] && board[i + 1][j] == 'O')
            dfs(board, notFlip, i + 1, j);
        
        // Go up
        if (i - 1 >= 0 && !notFlip[i - 1][j] && board[i - 1][j] == 'O')
            dfs(board, notFlip, i - 1, j);
        
        // Go left
        if (j - 1 >= 0 && !notFlip[i][j - 1] && board[i][j - 1] == 'O')
            dfs(board, notFlip, i, j - 1);
        
        // Go down
        if (j + 1 < n && !notFlip[i][j + 1] && board[i][j + 1] == 'O')
            dfs(board, notFlip, i, j + 1);
    }
}